packages/core/database/lib/query/helpers/streams/readable.js

```diff
@@ -0,0 +1,176 @@
+    // Stream params
+    this._batchSize = batchSize;
+    this._mapResults = mapResults;
+
+    // States
+    this[knexPerformingQuery] = false;
+  }
+
+  _destroy(err, cb) {
+    // If the stream is destroyed while a query is being made, then wait for a
+    // kQueryDone event to be emitted before actually destroying the stream
+    if (this[knexPerformingQuery]) {
+      this.once(knexQueryDone, (er) => cb(err || er));
+    } else {
+      cb(err);
+    }
+  }
+
+  /**
+   * Custom ._read() implementation
+   *
+   *  NOTE: Here "size" means the number of entities to be read from the database.
+   *  Not the actual byte size, as it would means that we need to return partial entities.
+   *
+   * @param {number} size
+   */
+  async _read(size) {
+    const query = this._query;
+
+    // Remove the original offset & limit properties from the query
+    // Theoretically, they would be replaced by calling them again, but this is just to be sure
+    query.clear('limit').clear('offset');
+
+    // Define the maximum read size based on the limit and the requested size
+    // NOTE: size is equal to _batchSize by default. Since we want to allow customizing it on
+    // the fly, we need to use its value instead of batchSize when computing the maxReadSize value
+    const maxReadSize =
+      // if no limit is defined in the query, use the given size,
+      // otherwise, use the smallest value between the two
+      this._limit === null ? size : Math.min(size, this._limit);
+
+    // Compute the limit for the next query
+    const limit =
+      // If a limit is defined
+      this._limit !== null &&
+      // And reading `maxReadSize` would fetch too many entities (> _limit)
+      this._fetched + maxReadSize > this._limit
+        ? // Then adjust the limit so that it only get the remaining entities
+          this._limit - this._fetched
+        : // Else, use the max read size
+          maxReadSize;
+
+    // If we don't have anything left to read (_limit === _fetched),
+    // don't bother making the query and end the stream by pushing null
+    if (limit <= 0) {
+      this.push(null);
+      return;
+    }
+
+    // Compute the offset (base offset + number of entities already fetched)
+    const offset = this._offset + this._fetched;
+
+    // Update the query with the new values (offset + limit)
+    query.offset(offset).limit(limit);
+
+    // Lock the ._destroy()
+    this[knexPerformingQuery] = true;
+
+    let results;
+    let count;
+    let err;
+
+    try {
+      // Execute the query and store the results & count
+      results = await query;
+
+      const { populate } = this._qb.state;
+
+      // Apply populate if needed
+      if (populate && results) {
+        await applyPopulate(results, populate, { qb: this._qb, uid: this._uid, db: this._db });
+      }
+
+      // Map results if asked to
+      if (this._mapResults && results) {
+        results = fromRow(this._meta, results);
+      }
+
+      count = results.length;
```

Comment:
I was going to say we should check if(results) here too, but we should just have a check for if(!results) at the top so we don't have to check it everywhere.

Reply:
Well according to knex typings, the return value will always be an array. If the table doesn't exist (or if the query is invalid for another reason). It'll throw an error.
So I think we can just remove the checks around `results`

Reply:
Sounds good to me